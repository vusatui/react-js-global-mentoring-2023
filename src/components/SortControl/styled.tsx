"use client";

import styled from "styled-components";

import { Listbox } from '@headlessui/react'

import ArrowIcon from "../icons/ArrowIcon";

export const StyledSortControl = styled.div`
  display: flex;
  flex-flow: row nowrap;
`;

export const StyledSortControlLabel = styled.label`
  margin-right: 31px;
  font-family: 'Montserrat',sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 20px;
  letter-spacing: 0.888889px;
  text-transform: uppercase;
  color: #FFFFFF;
  mix-blend-mode: normal;
  opacity: 0.6;
`;

export const StyledSortControlSelectInputWrapper = styled.div`
  position: relative;
`;

export const StyledSortControlSelectInput = styled(Listbox)`
  
`;

export const StyledSortControlSelectValue = styled.span`
  margin-right: 13px;
  font-family: 'Montserrat',sans-serif;
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  line-height: 20px;
  letter-spacing: 0.888889px;
  text-transform: uppercase;
  color: #FFFFFF;
`;

export const StyledSortControlSelectIcon = styled(ArrowIcon)<{ isRotated: boolean }>`
  width: 12px;
  transform-origin: center;
  ${({ isRotated }) => isRotated && `transform: rotate(180deg);`}
`;

export const StyledSortControlSelectOptions = styled(Listbox.Options)`
  margin-top: 10px;
  padding: 20px;
  background: rgba(35, 35, 35, 0.918051);
  box-shadow: 
      0 0 2px rgba(0, 0, 0, 0.1),
      0 2px 10px rgba(0, 0, 0, 0.1), 
      0 10px 20px rgba(0, 0, 0, 0.1), 
      0 10px 50px rgba(0, 0, 0, 0.1);
  backdrop-filter: blur(13.5914px);
  border-radius: 4px;
  
  * {
    padding-bottom: 18px;
  }
  
  &:last-child {
    padding-bottom: 0;
  }
`;

export const StyledSortControlSelectOption = styled(Listbox.Option)`
  font-family: 'Montserrat',sans-serif;
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  line-height: 20px;
  letter-spacing: 0.888889px;
  text-transform: uppercase;
  color: #FFFFFF;
  cursor: pointer;
`;

export const StyledSortControlSelectButton = styled(Listbox.Button).attrs({ "data-testid": "button" })`
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: center;
  width: 165px;
  background: transparent;
  border: none;
  outline: none;
  appearance: none;
  padding: 0;
  cursor: pointer;
`;