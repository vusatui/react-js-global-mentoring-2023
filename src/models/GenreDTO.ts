export interface GenreDTO {
    name: string
}