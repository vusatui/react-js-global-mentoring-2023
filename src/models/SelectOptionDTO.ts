export interface SelectOptionDTO<T> {
    id: string
    label: string
    value: T
}