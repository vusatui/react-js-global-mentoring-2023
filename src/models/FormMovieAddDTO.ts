import {GenreDTO} from "./GenreDTO";

export interface FormMovieAddDTO {
    title: string
    movieUrl: string
    genre: GenreDTO[]
    overview: string
    runtime: string
    rating: string
    releaseDate: string
}