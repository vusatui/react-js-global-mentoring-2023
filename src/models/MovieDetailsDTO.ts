import {GenreDTO} from "./GenreDTO";

export interface MovieDetailsDTO {
    imageUrl: string
    title: string
    year: string
    genres: GenreDTO[]
    duration: string
    rating: number
    description: string
}