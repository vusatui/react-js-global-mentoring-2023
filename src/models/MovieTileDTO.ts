import {GenreDTO} from "./GenreDTO";

export interface MovieTileDTO {
    imageUrl: string
    title: string
    year: string
    genres: GenreDTO[]
}