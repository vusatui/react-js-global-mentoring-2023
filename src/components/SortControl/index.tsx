"use client";

import {useMemo} from "react";

import {SelectOptionDTO} from "../../models/SelectOptionDTO";

import {
    StyledSortControl,
    StyledSortControlLabel, StyledSortControlSelectButton, StyledSortControlSelectIcon,
    StyledSortControlSelectInput, StyledSortControlSelectInputWrapper, StyledSortControlSelectOption,
    StyledSortControlSelectOptions, StyledSortControlSelectValue
} from "./styled";


interface SortControlProps {
    className?: string
    options: SelectOptionDTO<string>[]
    value: string
    onSelect: (value: string) => void
}

const SortControl = ({
    className,
    options,
    value,
    onSelect,
}: SortControlProps) => {
    const selectedLabel = useMemo<string|undefined>(
        () => options.filter(option => option.value === value).pop()?.label,
        [options, value],
    );

    const handleSelect = (value: string) => onSelect(value);

    return (
        <StyledSortControl className={className}>
            <StyledSortControlLabel>
                Sort by
            </StyledSortControlLabel>
            <StyledSortControlSelectInputWrapper>
                <StyledSortControlSelectInput
                    value={value}
                    onChange={handleSelect}
                >
                    {({ open }) => (
                        <>
                            <StyledSortControlSelectButton>
                                <StyledSortControlSelectValue>
                                    {selectedLabel}
                                </StyledSortControlSelectValue>
                                <StyledSortControlSelectIcon isRotated={open} />
                            </StyledSortControlSelectButton>

                            <StyledSortControlSelectOptions>
                                {options.map(option => (
                                    <StyledSortControlSelectOption
                                        key={option.id}
                                        value={option.value}
                                    >
                                        {option.label}
                                    </StyledSortControlSelectOption>
                                ))}
                            </StyledSortControlSelectOptions>
                        </>
                    )}
                </StyledSortControlSelectInput>
            </StyledSortControlSelectInputWrapper>
        </StyledSortControl>
    );
};

export default SortControl;