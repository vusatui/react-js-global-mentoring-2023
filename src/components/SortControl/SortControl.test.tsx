import {render} from "@testing-library/react";

import SortControl from "./index";

import {SelectOptionDTO} from "../../models/SelectOptionDTO";

const options: SelectOptionDTO<string>[] = [
    { id: "1", label: "Release Date", value: "Release Date" },
    { id: "2", label: "Title", value: "Title" },
];

const mockHandleSelect = jest.fn();

const Component = () => (<SortControl options={options} value={"Release Date"} onSelect={mockHandleSelect} />)

describe("SortControl", () => {

    it("Should renders correctly", () => {
        const { asFragment } = render(<Component />);
        expect(asFragment()).toMatchSnapshot();
    });

});