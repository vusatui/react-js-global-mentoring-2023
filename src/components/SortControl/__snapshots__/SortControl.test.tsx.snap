// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortControl Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-beqWaB kBGGqf"
  >
    <label
      class="sc-gueYoa kZFTxO"
    >
      Sort by
    </label>
    <div
      class="sc-dmqHEX hutmsU"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="sc-fsQiph bEzjAP"
        data-headlessui-state=""
        data-testid="button"
        id="headlessui-listbox-button-:r0:"
        type="button"
      >
        <span
          class="sc-eDDNvR fzsmLY"
        >
          Release Date
        </span>
        <svg
          class="sc-jTrPJq fKMRYO"
          fill="none"
          height="12"
          viewBox="0 0 18 12"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M1.13423 0C0.286533 0 -0.176668 0.988581 0.365828 1.63996L8.22615 11.0779C8.6258 11.5578 9.36265 11.558 9.76259 11.0784L17.6322 1.64041C18.1752 0.989135 17.7121 0 16.8642 0H1.13423Z"
            fill="#F65261"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
